src/librustc/ty/layout.rs

```diff
@@ -944,33 +940,33 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
 
-                            let mut abi = match st[i].abi {
-                                Abi::Scalar(_) => Abi::Scalar(niche_scalar.clone()),
-                                Abi::ScalarPair(ref first, ref second) => {
-                                    // We need to use scalar_unit to reset the
-                                    // valid range to the maximal one for that
-                                    // primitive, because only the niche is
-                                    // guaranteed to be initialised, not the
-                                    // other primitive.
-                                    if offset.bytes() == 0 {
-                                        Abi::ScalarPair(
-                                            niche_scalar.clone(),
-                                            scalar_unit(second.value),
-                                        )
-                                    } else {
-                                        Abi::ScalarPair(
-                                            scalar_unit(first.value),
-                                            niche_scalar.clone(),
-                                        )
+                            let abi = if st.iter().all(|v| v.abi.is_uninhabited()) {
+                                Abi::Uninhabited
+                            } else {
+                                match st[i].abi {
+                                    Abi::Scalar(_) => Abi::Scalar(niche_scalar.clone()),
+                                    Abi::ScalarPair(ref first, ref second) => {
+                                        // We need to use scalar_unit to reset the
+                                        // valid range to the maximal one for that
+                                        // primitive, because only the niche is
+                                        // guaranteed to be initialised, not the
+                                        // other primitive.
+                                        if offset.bytes() == 0 {
+                                            Abi::ScalarPair(
+                                                niche_scalar.clone(),
+                                                scalar_unit(second.value),
+                                            )
+                                        } else {
+                                            Abi::ScalarPair(
+                                                scalar_unit(first.value),
+                                                niche_scalar.clone(),
+                                            )
+                                        }
                                     }
+                                    _ => Abi::Aggregate { sized: true },
                                 }
-                                _ => Abi::Aggregate { sized: true },
                             };
 
-                            if st.iter().all(|v| v.abi.is_uninhabited()) {
```

Comment:
Clippy managed to identify this refactoring?! That's amazing -- I'm very surprised if so, given the amount of function calls etc. in the above block. It looks like it's safe to do in this case though -- no side effects above that don't happen now.

Reply:
````
warning: `if _ { .. } else { .. }` is an expression
   --> src/librustc/ty/layout.rs:947:29
    |
947 | / ...                   let mut abi = match st[i].abi {
948 | | ...                       Abi::Scalar(_) => Abi::Scalar(niche_scalar.clone()),
949 | | ...                       Abi::ScalarPair(ref first, ref second) => {
950 | | ...                           // We need to use scalar_unit to reset the
...   |
971 | | ...                       abi = Abi::Uninhabited;
972 | | ...                   }
    | |_______________________^
    |
    = note: you might not need `mut` at all
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_let_if_seq
help: it is more idiomatic to write
    |
947 |                             let <mut> abi = if st.iter().all(|v| v.abi.is_uninhabited()) { Abi::Uninhabited } else { match st[i].abi {
948 |                                 Abi::Scalar(_) => Abi::Scalar(niche_scalar.clone()),
949 |                                 Abi::ScalarPair(ref first, ref second) => {
950 |                                     // We need to use scalar_unit to reset the
951 |                                     // valid range to the maximal one for that
952 |                                     // primitive, because only the niche is
  ...
````

:)